src/shims/unix/fs.rs

```diff
@@ -81,6 +81,48 @@ impl FileDescription for FileHandle {
         }
     }
 
+    fn flock<'tcx>(
+        &self,
+        communicate_allowed: bool,
+        op: FlockOp,
+    ) -> InterpResult<'tcx, io::Result<()>> {
+        assert!(communicate_allowed, "isolation should have prevented even opening a file");
+        #[cfg(unix)]
+        {
+            use std::os::fd::AsRawFd;
+
+            // Transform `op` into host's flags
+            let mut host_op = 0;
+            if op.shared {
+                host_op |= libc::LOCK_SH;
+            }
+            if op.exclusive {
+                host_op |= libc::LOCK_EX;
+            }
+            if op.nonblocking {
+                host_op |= libc::LOCK_NB;
+            }
+            if op.unlock {
+                host_op |= libc::LOCK_UN;
+            }
+
+            let fd = self.file.as_raw_fd();
+            let ret = unsafe { libc::flock(fd, host_op) };
+            let res = match ret {
+                0 => Ok(()),
+                -1 => Err(io::Error::last_os_error()),
+                ret => panic!("Unexpected return value from flock: {ret}"),
+            };
+            Ok(res)
+        }
+
+        #[cfg(not(unix))]
+        {
+            let _ = op;
+            throw_unsup_format!("flock emulation is not supported on non-UNIX hosts");
```

Comment:
It may be possible to emulate `flock` on Windows using [`LockFileEx`](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-lockfileex), but I am not familiar with this API, so I do not plan to implement it.

Reply:
Instead of doing this work ourselves in miri, is https://crates.io/crates/fs4 tunable enough to emulate this correctly?

Reply:
I think it's a small enough code to include it into Miri directly. I will try to write it based on the `fs4` code.